src/plugins/nvme/nvme-error.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (C) 2014-2021 Red Hat, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Tomas Bzatek <tbzatek@redhat.com>
+ */
+
+#include <glib.h>
+#include <string.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <malloc.h>
+
+#include <libnvme.h>
+#include <uuid/uuid.h>
+
+#include <blockdev/utils.h>
+#include <check_deps.h>
+#include "nvme.h"
+#include "nvme-private.h"
+
+
+/**
+ * bd_nvme_error_quark: (skip)
+ */
+GQuark bd_nvme_error_quark (void)
+{
+    return g_quark_from_static_string ("g-bd-nvme-error-quark");
+}
+
+
+
+static inline const gchar * _nvme_generic_status_to_string (guint status)
+{
+    /*
+     * Generic Command Status Codes:
+     */
+    switch (status) {
+    case NVME_SC_INVALID_OPCODE:       /* 0x1 */
+        return "INVALID_OPCODE: The associated command opcode field is not valid";
```

Comment:
Do you think it would possible to get these error strings directly from libnvme? It looks like a lot of internal stuff to have in libblockdev. We did something similar for mount errors in the fs plugin and kzak agreed to add a "get error string" function for us.

Reply:
Posted upstream: https://github.com/linux-nvme/libnvme/pull/19